Fix wasm->JS when useJIT=0

danlliu opened the pull request:
#### 602054ed517608e2babeb5091d17be16abb0f155
<pre>
Fix wasm-&gt;JS when useJIT=0
<a href="https://bugs.webkit.org/show_bug.cgi?id=281512">https://bugs.webkit.org/show_bug.cgi?id=281512</a>
<a href="https://rdar.apple.com/138422560">rdar://138422560</a>

Reviewed by Yusuke Suzuki.

This patch fixes <a href="https://github.com/WebKit/WebKit/pull/35232">https://github.com/WebKit/WebKit/pull/35232</a>,
and consolidates those patches together. The major code changes were made by
Justin Michaud.

    1) Generally clean up this code, and remove magic numbers
    2) Ensure that the Callee stack slot is always a boxed callee; This did
        not seem to break any tests somehow, but it would be wrong when we
        try to unwind.
    3) Abstract both an imported JS or wasm function, or a function in a table
        into a WasmCallableFunction, and pass that from LLint to LLInt calls.
        The JIT case does not need this.

    This should fix the last remaining ARMv7 test failure, although this feature
    in particular is broken on all platforms.

This patch cleans up the ~80 test failures in `--jitless-wasm` test mode.
Some of these failures were due to GC tests that should not be running, which have
been disabled. The key fixes involve:

    - Fixing where wasm_to_js_wrapper_entry loads its WasmCallableFunction*
    - Fixing a conflict in tail calls where the WasmCallableFunction* would get
      overwritten by another register being saved in the CodeBlock slot
    - Update the WasmToJSExit operations to use the WasmCallableFunction*, and
      fixing some cases where properties of the WasmCallableFunction wouldn&apos;t
      be set properly.

* JSTests/wasm/gc/bug254413.js:
* JSTests/wasm/gc/bug254414.js:
* JSTests/wasm/gc/bug258127.js:
* JSTests/wasm/gc/bug258128.js:
* JSTests/wasm/stress/simd-global.js:
* JSTests/wasm/stress/try-and-block-with-v128-results.js:
* JSTests/wasm/stress/try-table-malformed-catch-label.js:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitTailCall):
(JSC::Wasm::BBQJITImpl::BBQJIT::addCall):
(JSC::Wasm::BBQJITImpl::BBQJIT::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmBinding.cpp:
(JSC::Wasm::wasmToWasm):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::WasmCallableFunction::offsetOfEntrypointLoadLocation):
(JSC::Wasm::WasmCallableFunction::offsetOfBoxedWasmCalleeLoadLocation):
(JSC::Wasm::WasmToWasmImportableFunction::offsetOfSignatureIndex):
(JSC::Wasm::WasmToWasmImportableFunction::offsetOfEntrypointLoadLocation): Deleted.
(JSC::Wasm::WasmToWasmImportableFunction::offsetOfBoxedWasmCalleeLoadLocation): Deleted.
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::doWasmCall):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::doWasmCall):
(JSC::LLInt::doWasmCallIndirect):
(JSC::LLInt::doWasmCallRef):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::FuncRefTable::setFunction):
* Source/JavaScriptCore/wasm/WasmTable.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.cpp:
(JSC::JSWebAssemblyInstance::JSWebAssemblyInstance):
(JSC::JSWebAssemblyInstance::~JSWebAssemblyInstance):
(JSC::JSWebAssemblyInstance::finalizeCreation):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::WebAssemblyFunction::WebAssemblyFunction):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.cpp:
(JSC::WebAssemblyFunctionBase::WebAssemblyFunctionBase):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
(JSC::WebAssemblyFunctionBase::importableFunction const):
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeImports):
* Source/JavaScriptCore/wasm/js/WebAssemblyWrapperFunction.cpp:
(JSC::WebAssemblyWrapperFunction::WebAssemblyWrapperFunction):
(JSC::WebAssemblyWrapperFunction::create):

Canonical link: <a href="https://commits.webkit.org/288046@main">https://commits.webkit.org/288046@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e84094da5ea57e4a77d78818943f5fcfc4bd5896

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81702 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1228 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35656 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86246 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32697 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83808 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1261 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9048 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63747 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21471 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84772 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/914 "Found 1 new test failure: fast/forms/ios/file-upload-panel.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74356 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44033 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/813 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28530 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31151 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74682 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29132 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87685 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80756 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8942 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6351 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72081 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9127 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70175 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71311 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17773 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15397 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14312 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103167 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8893 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14425 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25057 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8734 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12257 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10542 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -276,12 +276,12 @@ macro restoreCalleeSavesUsedByWasm()
 
 macro restoreGPRsUsedByTailCall(gpr0, gpr1)
-    loadp CodeBlock[sp], gpr0
+    loadp ThisArgumentOffset[sp], gpr0
```
Is this change correct? Isn't it possible that this location gets clobbered during tail-call shuffling?

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
```diff
@@ -5130,7 +5130,9 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addCallRef(const TypeDefinition& origin
             m_jit.loadPtr(Address(calleePtr, WebAssemblyFunctionBase::offsetOfBoxedWasmCalleeLoadLocation()), calleeTmp);
+            auto branch = m_jit.branchTestPtr(ResultCondition::Zero, calleeTmp);
             m_jit.loadPtr(Address(calleeTmp), calleeTmp);
+            branch.link(&m_jit);
```
When does it happen?

danlliu replied:
This happens in cases where we create a WebAssemblyWrapperFunction; since that initializes our load location to 0, we need to branch. I've confirmed that adding these branches is neutral on perf.

Constellation replied:
Does it happen without `useJIT=0`? Or is it happening only when `useJIT=0`? If it is latter, do you know why this happens only when `useJIT=0`?

danlliu replied:
This happens without `useJIT=0` as well; it's just a byproduct of the change to bookkeeping Callee load locations.

Constellation replied:
For `WebAssemblyWrapperFunction`, can we put a thunk calling a function instead of branching? I'm concerning about inserting null-check for each CallRef case.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -5668,9 +5668,25 @@ auto OMGIRGenerator::addCallRef(const TypeDefinition& originalSignature, Argumen
+    m_currentBlock = continuation;
+    Value* calleeCallee = m_currentBlock->appendNew<Value>(m_proc, Phi, pointerType(), origin());
+    nonzeroUpsilon->setPhi(calleeCallee);
+    zeroUpsilon->setPhi(calleeCallee);
```
Ditto. I wonder if we can put a thunk instead of zero, so that this check is not necessary.

Constellation reviewed (changes requested):
Commented

Constellation commented on JSTests/wasm/stress/wasm-js-call-many-return-types-on-stack-no-args.js:
```diff
@@ -35,6 +35,7 @@ async function buildWat(types) {
 async function test() {
     await Promise.all([
         buildWat(["i32", "i32", "f32", "i32"]),
+        /*
```
What is the reason of disabling these tests?

danlliu committed:
Fix wasm->JS when useJIT=0
https://bugs.webkit.org/show_bug.cgi?id=281512
rdar://138422560

Reviewed by Yusuke Suzuki.

This patch fixes https://github.com/WebKit/WebKit/pull/35232,
and consolidates those patches together. The major code changes were made by
Justin Michaud.

    1) Generally clean up this code, and remove magic numbers
    2) Ensure that the Callee stack slot is always a boxed callee; This did
        not seem to break any tests somehow, but it would be wrong when we
        try to unwind.
    3) Abstract both an imported JS or wasm function, or a function in a table
        into a WasmCallableFunction, and pass that from LLint to LLInt calls.
        The JIT case does not need this.

    This should fix the last remaining ARMv7 test failure, although this feature
    in particular is broken on all platforms.

This patch cleans up the ~80 test failures in `--jitless-wasm` test mode.
Some of these failures were due to GC tests that should not be running, which have
been disabled. The key fixes involve:

    - Fixing where wasm_to_js_wrapper_entry loads its WasmCallableFunction*
    - Fixing a conflict in tail calls where the WasmCallableFunction* would get
      overwritten by another register being saved in the CodeBlock slot
    - Update the WasmToJSExit operations to use the WasmCallableFunction*, and
      fixing some cases where properties of the WasmCallableFunction wouldn't
      be set properly.

* JSTests/wasm/gc/bug254413.js:
* JSTests/wasm/gc/bug254414.js:
* JSTests/wasm/gc/bug258127.js:
* JSTests/wasm/gc/bug258128.js:
* JSTests/wasm/stress/simd-global.js:
* JSTests/wasm/stress/try-and-block-with-v128-results.js:
* JSTests/wasm/stress/try-table-malformed-catch-label.js:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitTailCall):
(JSC::Wasm::BBQJITImpl::BBQJIT::addCall):
(JSC::Wasm::BBQJITImpl::BBQJIT::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmBinding.cpp:
(JSC::Wasm::wasmToWasm):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::WasmCallableFunction::offsetOfEntrypointLoadLocation):
(JSC::Wasm::WasmCallableFunction::offsetOfBoxedWasmCalleeLoadLocation):
(JSC::Wasm::WasmToWasmImportableFunction::offsetOfSignatureIndex):
(JSC::Wasm::WasmToWasmImportableFunction::offsetOfEntrypointLoadLocation): Deleted.
(JSC::Wasm::WasmToWasmImportableFunction::offsetOfBoxedWasmCalleeLoadLocation): Deleted.
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::doWasmCall):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::doWasmCall):
(JSC::LLInt::doWasmCallIndirect):
(JSC::LLInt::doWasmCallRef):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::FuncRefTable::setFunction):
* Source/JavaScriptCore/wasm/WasmTable.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.cpp:
(JSC::JSWebAssemblyInstance::JSWebAssemblyInstance):
(JSC::JSWebAssemblyInstance::~JSWebAssemblyInstance):
(JSC::JSWebAssemblyInstance::finalizeCreation):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::WebAssemblyFunction::WebAssemblyFunction):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.cpp:
(JSC::WebAssemblyFunctionBase::WebAssemblyFunctionBase):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
(JSC::WebAssemblyFunctionBase::importableFunction const):
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeImports):
* Source/JavaScriptCore/wasm/js/WebAssemblyWrapperFunction.cpp:
(JSC::WebAssemblyWrapperFunction::WebAssemblyWrapperFunction):
(JSC::WebAssemblyWrapperFunction::create):

Canonical link: https://commits.webkit.org/288046@main
